### PR TITLE
Cleanup and modernize Optional.h

### PR DESCRIPTION
**Pull request by heitorschueroff**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64092
* #64084
* __->__ #64020

### CODE REVIEW

File: c10/util/Optional.h [hunk 1]
```diff
@@ -426,45 +407,44 @@ struct trivially_copyable_optimization_optional_base {
...
+    ArrayRefT value_;
 
-    constexpr void setUninitialized() noexcept {
-      uninitialized_.p = nullptr;
-      uninitialized_.sz = 1;
-    }
+    constexpr storage() noexcept : uninitialized_() {}
 
-    explicit constexpr storage(ArrayRefT& v) : value_(v) {}
+    explicit constexpr storage(const ArrayRefT& v) : value_(v) {}
```
Comment (ezyang) [inline c10/util/Optional.h:426]: Hmm, you sure?

**Comment by ezyang** [comment]
These seems OK I guess, deferring to @swolchok

File: c10/util/Optional.h [hunk 2]
```diff
@@ -426,45 +407,44 @@ struct trivially_copyable_optimization_optional_base {
...
     struct raw {
       // ArrayRef has the invariant that if Data is nullptr then
       // Length must be zero, so this is an unused bit pattern.
-      const void* p = nullptr;
-      size_t sz = 1;
-    } uninitialized_{};
-    ArrayRefT value_;
+      const T* p{nullptr};
+      size_type sz{1};
+    } uninitialized_;
```
Comment (swolchok) [inline c10/util/Optional.h:420]: it's silly, but I think removing the braces here will cause clang-tidy to complain. If it doesn't, great!

Do the braces vs = signs on the preceding two lines make a difference or is it just a style thing?

**Review by swolchok — COMMENTED** [review]
seems fine; I'm a little bit reluctant to make a bunch of style changes to this code because it makes merging any upstream fixes that much harder, but perhaps we've passed that point anyway.

**Pull request closed without merging**